### ISSUE TITLE
Add Provider.orElse(Callable)

### ISSUE THREAD

**Issue by lptr** [gradle/gradle#29358]
When the user wants to fall back to a calculated value in case a `Provider` has no value associated with it, currently they have the following options:

1) Rely on `Project` to create the provider, which means we are introducing a dependence on the whole of the project state that is unnecessary:

```java
public DependencyReportTask() {
    getConfigurations().convention(
        getDefaultConfigurationName()
            .map(this::lookupConfigurationByName)
            .orElse(
                getProject().provider(this::getNonDeprecatedTaskConfigurations)
            )
    );
}
```

2) Inject `ProviderFactory` which adds a lot of boilerplate:

```java
@Inject
public DependencyReportTask(ProviderFactory providerFactory) {
    getConfigurations().convention(
        getDefaultConfigurationName()
            .map(this::lookupConfigurationByName)
            .orElse(
                // Use ProviderFactory to create a provider
                providerFactory.provider(this::getNonDeprecatedTaskConfigurations)
            )
    );
}
```

This would be a lot nicer:

```
public DependencyReportTask() {
    getConfigurations().convention(
        getDefaultConfigurationName()
            .map(this::lookupConfigurationByName)
            .orElse(
                // Just pass a Callable
                this::getNonDeprecatedTaskConfigurations
            )
    );
}
```

And given that `ProviderFactory.provider(Callable)` does nothing else, just wraps the `Callable` in a `DefaultProvider`, we could easily add this method.

**Comment by mlopatkin** [comment]
Maybe a static `provider(Callable)` method would be enough?

**Comment by adammurdoch** [comment]
I think we should try to move away from using `Callable` to provide calculated values, as it is very error prone. It is easy for build dependencies and provenance to go missing, and we've seen this cause strange, difficult to diagnose problems in practice many times. Of course, there are ways to address this, but in general I think it might be cheaper overall to not use `Callable`.

So, in the example above, I would have `getNonDeprecatedTaskConfigurations()` return a Provider and use `orElse(getNonDeprecatedTaskConfigurations())`.

**Comment by lptr** [comment]
I agree that in this case it would be better to have `getNonDeprecatedTaskConfigurations()` return a `Provider` -- that way we could track where the value comes from all the way back to the configurations themselves.
